Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp

```diff
@@ -0,0 +1,775 @@
+      {
+        if (NetRef[i].TcpBuffers[l].used)
+        {
+          if (NetRef[i].TcpBuffers[l].seq_id < Common::swap32(tcpdata->ack_num))
+          {
+            NetRef[i].TcpBuffers[l].used = false;  // confirmed data received
+            if (!NetRef[i].ready && !NetRef[i].TcpBuffers[0].used)
+              NetRef[i].ready = true;
+          }
+        }
+      }
+    }
+  }
+}
+
+
+bool CEXIETHERNET::BuiltInBBAInterface::SendFrame(const u8* frame, u32 size)
+{
+  net_hw_lvl* hwdata;
+  net_ipv4_lvl* ipdata;
+  net_udp_lvl* udpdata;
+  net_tcp_lvl* tcpdata;
+  int offset = 0;
+  int i;
+  sf::IpAddress target;
+  mtx.lock();
+  isSent = true;
+  memmove(m_out_frame, frame, size);
+
+  // handle the packet data
+  hwdata = (net_hw_lvl*)&m_out_frame[0];
+  if (hwdata->protocol == 0x08)
+  {
+    // IP sub
+    ipdata = (net_ipv4_lvl*)&m_out_frame[14];
+    switch (ipdata->protocol)
+    {
+    case 17:  // udp
+      udpdata = (net_udp_lvl*)&m_out_frame[14 + ((ipdata->header & 0xf) * 4)];
+      offset = 14 + ((ipdata->header & 0xf) * 4) + 8;
+      if (Common::swap16(udpdata->dest_port) == 67)
+      {
+        net_dhcp* request = (net_dhcp*)&m_out_frame[offset];
+        HandleDHCP(hwdata, udpdata, request);
+      }
+      else
+      {
+        if (ipdata->dest_ip == 0)
+          ipdata->dest_ip = 0x0101a8c0;
+        // dns request
+        i = GetAvaibleSlot(udpdata->dest_port);
+        if (NetRef[i].ip == 0)
+        {
+          NetRef[i].ip = ipdata->dest_ip;  // change for ip
+          NetRef[i].local = udpdata->src_port;
+          NetRef[i].remote = udpdata->dest_port;
+          NetRef[i].type = 17;
+          udp_socket[i].setBlocking(false);
+          if (udp_socket[i].bind(sf::Socket::AnyPort) != sf::Socket::Done)
+          {
+            ERROR_LOG_FMT(SP1, "Couldn't open UDP socket");
+            mtx.unlock();
+            return false;
+          }
+        }
+        if (Common::swap16(udpdata->dest_port) == 53)
+        {
+          target = sf::IpAddress(m_dns_ip.c_str());  // dns server ip
+        }
+        else
+        {
+          target = sf::IpAddress(Common::swap32(ipdata->dest_ip));
+        }
+
+        udp_socket[i].send(&m_out_frame[offset], Common::swap16(udpdata->size) - 8, target,
+                           Common::swap16(udpdata->dest_port));
+      }
+
+      break;
+
+    case 6:  // tcp
+      tcpdata = (net_tcp_lvl*)&m_out_frame[14 + ((ipdata->header & 0xf) * 4)];
+      offset = 14 + ((ipdata->header & 0xf) * 4) + (((tcpdata->flag_length >> 4) & 0xf) * 4);
+      HandleTCPFrame(hwdata, ipdata, tcpdata, &m_out_frame[offset]);
+      break;
+    }
+  }
+  if (hwdata->protocol == 0x608)  // arp
+  {
+    net_arp_lvl* arpdata = (net_arp_lvl*)&m_out_frame[14];
+    HandleARP(hwdata, arpdata);
+  }
+
+  m_eth_ref->SendComplete();
+  mtx.unlock();
+  return true;
+}
+
+void CEXIETHERNET::BuiltInBBAInterface::ReadThreadHandler(CEXIETHERNET::BuiltInBBAInterface* self)
```

Comment:
You might want to split the function in smaller ones as this one is overly complicated.